File: app/src/main/java/me/lukeforit/spaceofaday/service/FetchApodFirebaseJobService.java
package me.lukeforit.spaceofaday.service;

import com.firebase.jobdispatcher.JobParameters;
import com.firebase.jobdispatcher.JobService;

import javax.inject.Inject;

import dagger.android.AndroidInjection;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;
import me.lukeforit.spaceofaday.common.Utils;
import me.lukeforit.spaceofaday.data.model.Apod;
import me.lukeforit.spaceofaday.data.source.SpaceRepository;

public class FetchApodFirebaseJobService extends JobService {

    @Inject
    SpaceRepository repository;

    private CompositeDisposable disposable = new CompositeDisposable();

    @Override
    public void onCreate() {
        AndroidInjection.inject(this);
        super.onCreate();
    }

    @Override
    public boolean onStartJob(final JobParameters job) {
        disposable.add(
                repository.fetchApod(Utils.getDefaultDateAsString())
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe(new Consumer<Apod>() {
                            @Override
                            public void accept(Apod apod) {
                                jobFinished(job, false);
                                //TODO notify about success
                            }
                        }, new Consumer<Throwable>() {
                            @Override
                            public void accept(Throwable throwable) {
                                //TODO log error
                                jobFinished(job, true);
                            }
                        })
        );
        return true;
    }

    @Override
    public boolean onStopJob(JobParameters job) {
        disposable.clear();
        return true;
    }
}

File: app/src/main/java/me/lukeforit/spaceofaday/service/FetchApodFirebaseJobService.kt
package me.lukeforit.spaceofaday.service

import com.firebase.jobdispatcher.JobParameters
import com.firebase.jobdispatcher.JobService
import dagger.android.AndroidInjection
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import me.lukeforit.spaceofaday.common.Utils
import me.lukeforit.spaceofaday.data.source.SpaceRepository
import me.lukeforit.spaceofaday.ui.widget.ApodWidget
import javax.inject.Inject

class FetchApodFirebaseJobService : JobService() {

    @Inject
    internal var repository: SpaceRepository? = null

    private val disposable = CompositeDisposable()

    override fun onCreate() {
        AndroidInjection.inject(this)
        super.onCreate()
    }

    override fun onStartJob(job: JobParameters): Boolean {
        disposable.add(
                repository!!.fetchApod(Utils.defaultDateAsString)
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe({
                            jobFinished(job, false)
                            ApodWidget.triggerWidgetUpdate(applicationContext)
                        }, {
                            //TODO log error
                            jobFinished(job, true)
                        })
        )
        return true
    }

    override fun onStopJob(job: JobParameters): Boolean {
        disposable.clear()
        return true
    }
}
